module/apmelasticsearch/clustername.go

```diff
@@ -0,0 +1,193 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package apmelasticsearch // import "go.elastic.co/apm/module/apmelasticsearch/v2"
+
+import (
+	"container/list"
+	"context"
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"sync"
+)
+
+// ClusterNameFunc is a function for fetching the Elasticsearch
+// cluster name for an HTTP request.
+type ClusterNameFunc func(*http.Response) string
+
+// WithClusterName returns a ClientOption which sets f as the
+// function to use to obtain the Elasticsearch cluster name for
+// recording on spans.
+func WithClusterName(f ClusterNameFunc) ClientOption {
+	if f == nil {
+		panic("f == nil")
+	}
+	return func(rt *roundTripper) {
+		rt.clusterNameFunc = f
+	}
+}
+
+// DefaultClusterName returns the default ClusterNameFunc implementation
+// used by WrapRoundTripper, if WithClusterName is not specified.
+//
+// DefaultClusterName first looks for the X-Found-Handling-Cluster response
+// header, using that if received; this is set by Elastic Cloud to the target
+// cluster name. If the response header is not received, DefaultClusterName
+// will make a query to "/_nodes/http" using the supplied RoundTripper, and
+// cache the result by the request URL's Host.
+//
+// Each invocation of DefaultClusterName will return a distinct ClusterNameFunc
+// with its own cached host-to-cluster-name mappings.
+func DefaultClusterName(rt http.RoundTripper) ClusterNameFunc {
+	client := &http.Client{Transport: rt}
+	queryClusterName := func(
+		ctx context.Context,
+		requestURL *url.URL,
+		header http.Header,
+	) (string, error) {
+		// Query /_nodes/http. This request is expected to fail for
+		// clients that are not authorized.
+		u := url.URL{
+			Scheme: requestURL.Scheme,
+			User:   requestURL.User,
+			Host:   requestURL.Host,
+			Path:   "/_nodes/http",
+		}
+
+		req, err := http.NewRequest(http.MethodGet, u.String(), nil)
```

Comment:
Should we use the context that is passed as an argument for the request?
```suggestion
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
```

Reply:
Yikes, we should indeed. Thanks for the catch!